File: docs/snippets/all/howto/component_mapping.py
"""
Demonstrates how to configure visualizer component mappings from blueprint.

⚠️TODO(#12600): The API for component mappings is still evolving, so this example may change in the future.
"""

from __future__ import annotations


import numpy as np
import pyarrow as pa
import rerun as rr
import rerun.blueprint as rrb
from rerun.blueprint.datatypes import ComponentSourceKind, VisualizerComponentMapping


# region: nested_struct
def make_sigmoid_struct_array(steps: int) -> pa.StructArray:
    """Creates a StructArray with a `values` field containing sigmoid data.

    Note: We intentionally use float32 here to demonstrate that the data will be
    automatically cast to the correct type (float64) when resolved by the visualizer.
    """
    x = np.arange(steps, dtype=np.float32) / 10.0
    sigmoid_values = 1.0 / (1.0 + np.exp(-(x - 3.0)))
    return pa.StructArray.from_arrays([pa.array(sigmoid_values, type=pa.float32())], names=["values"])


# endregion: nested_struct


rr.init("rerun_example_component_mapping", spawn=True)

# Send plot data using send_columns.
times = np.arange(64)
rr.send_columns(
    "plot",
    indexes=[rr.TimeColumn("step", sequence=times)],
    columns=[
        # Regular scalar batch with a sin.
        *rr.Scalars.columns(scalars=np.sin(times / 10.0)),
        # region: custom_data
        # Custom scalar batch with a cos using a custom component name.
        *rr.DynamicArchetype.columns(archetype="custom", components={"my_custom_scalar": np.cos(times / 10.0)}),
        # Nested custom scalar batch with a sigmoid inside a struct.
        *rr.DynamicArchetype.columns(
            archetype="custom", components={"my_nested_scalar": make_sigmoid_struct_array(64)}
        ),
        # endregion: custom_data
    ],
)

# Add a line series color to the store data.
rr.log("plot", rr.SeriesLines(colors=[255, 0, 0]), static=True)

# Create a blueprint with explicit component mappings
blueprint = rrb.Blueprint(
    rrb.TimeSeriesView(
        name="Component Mapping Demo",
        origin="/",
        # Set default color for series to blue.
        defaults=[rr.SeriesLines(colors=[0, 255, 0])],
        overrides={
            # Three line series visualizations for the "plot" entity:
            "plot": [
                # region: custom_value
                # Red sine:
                # * set the name via an override
                # * explicitly use the view's default for color
                # * everything else uses the automatic component mappings, so it will pick up scalars from the store.
                rr.SeriesLines(names="sine (store)").visualizer(
                    mappings=[
                        VisualizerComponentMapping(
                            target="SeriesLines:colors",
                            source_kind=ComponentSourceKind.Default,
                        ),
                    ]
                ),
                # endregion: custom_value
                # region: source_mapping
                # Green cosine:
                # * source scalars from the custom component "custom:my_custom_scalar"
                # * set the name via an override
                # * everything else uses the automatic component mappings, so it will pick up colors from the view default.
                # region: source_mapping
                rr.SeriesLines(names="cosine (custom)").visualizer(
                    mappings=[
                        # Map scalars to the custom component.
                        VisualizerComponentMapping(
                            target="Scalars:scalars",
                            source_kind=ComponentSourceKind.SourceComponent,
                            source_component="custom:my_custom_scalar",  # Map from custom component
                        ),
                    ]
                ),
                # endregion: source_mapping
                # region: selector_mapping
                # Blue sigmoid:
                # * source scalars from a nested struct using a selector to extract the "values" field
                # * set the name and an explicit blue color via overrides
                rr.SeriesLines(names="sigmoid (nested)", colors=[0, 0, 255]).visualizer(
                    mappings=[
                        VisualizerComponentMapping(
                            target="Scalars:scalars",
                            source_kind=ComponentSourceKind.SourceComponent,
                            source_component="custom:my_nested_scalar",
                            selector=".values",
                        ),
                    ]
                ),
                # endregion: selector_mapping
            ],
        },
    ),
)

rr.send_blueprint(blueprint)

File: docs/snippets/all/howto/component_mapping.rs
//! Demonstrates how to configure visualizer component mappings from blueprint.

use std::sync::Arc;

use rerun::AsComponents as _;
use rerun::blueprint::VisualizableArchetype as _;
use rerun::external::arrow::array::{Array, Float32Array, Float64Array, StructArray};
use rerun::external::arrow::datatypes::{DataType, Field};

// region: nested_struct
/// Creates a `StructArray` with a `values` field containing sigmoid data.
///
/// Note: We intentionally use `Float32` here to demonstrate that the data will be
/// automatically cast to the correct type (`Float64`) when resolved by the visualizer.
fn make_sigmoid_struct_array(steps: usize) -> StructArray {
    let sigmoid_values: Vec<f32> = (0..steps)
        .map(|step| {
            let x = step as f32 / 10.0;
            1.0 / (1.0 + (-(x - 3.0)).exp())
        })
        .collect();

    StructArray::from(vec![(
        Arc::new(Field::new("values", DataType::Float32, true)),
        Arc::new(Float32Array::from(sigmoid_values)) as Arc<dyn Array>,
    )])
}
// endregion: nested_struct

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_component_mapping").spawn()?;

    // Generate columns for regular scalars (sin)
    let sin = (0..64).map(|step| (step as f64 / 10.0).sin());
    let sin_columns = rerun::Scalars::new(sin).columns_of_unit_batches()?;

    // region: custom_data
    // Generate columns for custom component (cos)
    let cos = (0..64).map(|step| (step as f64 / 10.0).cos());
    let cos_array = Arc::new(cos.collect::<Float64Array>());
    let custom_columns = rerun::DynamicArchetype::new("custom")
        .with_component_from_data("my_custom_scalar", cos_array)
        .as_serialized_batches()
        .into_iter()
        .map(|batch| batch.column_of_unit_batches())
        .collect::<Result<Vec<_>, _>>()?;

    // Generate columns for nested custom component (sigmoid)
    let sigmoid_array = Arc::new(make_sigmoid_struct_array(64));
    let nested_columns = rerun::DynamicArchetype::new("custom")
        .with_component_from_data("my_nested_scalar", sigmoid_array)
        .as_serialized_batches()
        .into_iter()
        .map(|batch| batch.column_of_unit_batches())
        .collect::<Result<Vec<_>, _>>()?;
    // endregion: custom_data

    // Send plot data using send_columns.
    rec.send_columns(
        "plot",
        [rerun::TimeColumn::new_sequence("step", 0..64)],
        sin_columns.chain(custom_columns).chain(nested_columns),
    )?;

    // Add a line series color to the store data
    rec.log_static(
        "plot",
        &rerun::SeriesLines::new().with_colors([[255, 0, 0]]),
    )?;

    // Create a blueprint with explicit component mappings
    let blueprint = rerun::blueprint::Blueprint::new(
        rerun::blueprint::TimeSeriesView::new("Component Mapping Demo")
            .with_origin("/")
            // Set default color for series to green
            .with_defaults(&rerun::SeriesLines::new().with_colors([[0, 255, 0]]))
            .with_overrides(
                "plot",
                [
                    // region: custom_value
                    // Red sine:
                    // * set the name via an override
                    // * explicitly use the view's default for color
                    // * everything else uses the automatic component mappings, so it will pick up scalars from the store
                    rerun::SeriesLines::new()
                        .with_names(["sine (store)"])
                        .visualizer()
                        .with_mappings(vec![
                            rerun::blueprint::VisualizerComponentMapping::new_default(
                                rerun::SeriesLines::descriptor_colors().component,
                            )
                            .into(),
                        ]),
                    // endregion: custom_value
                    // region: source_mapping
                    // Green cosine:
                    // * source scalars from the custom component "plot:my_custom_scalar"
                    // * set the name via an override
                    // * everything else uses the automatic component mappings, so it will pick up colors from the view default
                    rerun::SeriesLines::new()
                        .with_names(["cosine (custom)"])
                        .visualizer()
                        .with_mappings(vec![
                            rerun::blueprint::VisualizerComponentMapping::new_source_component(
                                rerun::Scalars::descriptor_scalars().component,
                                "custom:my_custom_scalar",
                            )
                            .into(),
                        ]),
                    // endregion: source_mapping
                    // region: selector_mapping
                    // Blue sigmoid:
                    // * source scalars from a nested struct using a selector to extract the "values" field
                    // * set the name and an explicit blue color via overrides
                    rerun::SeriesLines::new()
                        .with_names(["sigmoid (nested)"])
                        .with_colors([[0, 0, 255]])
                        .visualizer()
                        .with_mappings(vec![
                            rerun::blueprint::VisualizerComponentMapping::new_source_component_with_selector(
                                "Scalars:scalars",
                                "custom:my_nested_scalar",
                                ".values",
                            )
                            .into(),
                        ]),
                    // endregion: selector_mapping
                ],
            ),
    );

    blueprint.send(&rec, rerun::blueprint::BlueprintActivation::default())?;

    Ok(())
}
